### PR TITLE
Reducing file IO and parsing in CLDR transformers

### PR DESCRIPTION
**Pull request by robertbastian**
This uses a `FrozenMap`, similar to the `elsa` crate but around a `BTreeMap` and with a `get_or_insert` method. This is used for `TimeZonesProvider`, which now doesn't have to clone the parsed files, and for `CommonDateProvider`, which now implements the date resource providers instead of existing three times inside each of them.

### CODE REVIEW

**Comment by Manishearth** [comment]
Could you upstream this into elsa instead? We'd happily accept a FrozenBTreeMap, we just haven't added one yet because nobody wanted one. Gives us less unsafe code in our tree, more in places that other people use.

**Comment by robertbastian** [comment]
Would you accept it as-is, with just the `get_or_insert` method, the error signature, and the `Box` as part of the type?

**Comment by Manishearth** [comment]
@robertbastian oops, missed this, sorry

It would have to use StableDeref instead of Box (see the other types). But otherwise it's fine if it just has one method for now. I can add more later.

**Comment by robertbastian** [comment]
https://github.com/Manishearth/elsa/pull/24

File: provider/cldr/src/transform/datetime/mod.rs [hunk 1]
```diff
@@ -2,8 +2,318 @@
...
+                    let langid = req
+                        .get_langid()
+                        .ok_or_else(|| DataErrorKind::NeedsLocale.with_req(<$marker>::KEY, req))?;
+                    let variant = req
+                        .options
+                        .variant
+                        .as_ref()
+                        .ok_or_else(|| DataErrorKind::NeedsVariant.with_req(<$marker>::KEY, req))?;
+
+                    let dates = self.data.get_or_insert(&req.options, || -> Result<_, DataError> {
```
Comment (Manishearth) [inline provider/cldr/src/transform/datetime/mod.rs:56]: Thinking about this a bit more, `get_or_insert` feels doable with just a get and insert (and has the same perf), I think I'd prefer to remove the API, thoughts?

Comment (Manishearth) [reply provider/cldr/src/transform/datetime/mod.rs:56]: @robertbastian try it out with elsa master and once you confirm it works I'll cut a new version

Comment (robertbastian) [reply provider/cldr/src/transform/datetime/mod.rs:56]: That was my one method :(. Doing separate `get`s and `inserts` adds control flow to the caller which makes the code more complex. Option also has `get_or_insert`, presumably for this reason.

Comment (robertbastian) [reply provider/cldr/src/transform/datetime/mod.rs]: I did update the code

Comment (Manishearth) [reply provider/cldr/src/transform/datetime/mod.rs]: Yeah I saw, I'll get you your version bump in a bit

File: provider/cldr/Cargo.toml [hunk 2]
```diff
@@ -59,6 +59,7 @@ icu_locid_macros = { version = "0.5", path = "../../components/locid/macros" }
 icu_provider_uprops = { version = "0.5", path = "../../provider/uprops", features = ["log"] }
 icu_properties = { version = "0.5", path = "../../components/properties", features = ["provider_serde", "std"] }
 zerovec = { version = "0.6", path = "../../utils/zerovec"}
+elsa = "1.7"
```
Comment (Manishearth) [inline provider/cldr/Cargo.toml:62]: note: needs approval from dependency cop @zbraniecki

File: provider/cldr/src/transform/datetime/mod.rs [hunk 3]
```diff
@@ -2,8 +2,320 @@
...
+                                    r
+                                }),
+                        );
+                    }
+                    Ok(Box::new(r.into_iter()))
+                }
+            }
+        )+
+
+        icu_provider::impl_dyn_provider!(CommonDateProvider, [$($marker),+,], SERDE_SE);
```
Comment (Manishearth) [inline provider/cldr/src/transform/datetime/mod.rs:130]: note: i think with the verification PR landed this needs to be a single invocation outside with the three markers

**Review by Manishearth — DISMISSED** [review]

File: provider/cldr/src/transform/datetime/mod.rs (hunk 3 above)
Comment (Manishearth) [reply provider/cldr/src/transform/datetime/mod.rs:130]: oh I see

Comment (Manishearth) [reply provider/cldr/src/transform/datetime/mod.rs:130]: It still needs `impl ConvertData` after you rebase, though

Comment (robertbastian) [reply provider/cldr/src/transform/datetime/mod.rs]: done

**Comment by Manishearth** [comment]
Can land after signoff from zibi

**Review by Manishearth — DISMISSED** [review]

**Review by Manishearth — APPROVED** [review]

**Comment by Manishearth** [comment]
Tests are failing though

**Review by Manishearth — APPROVED** [review]

File: provider/cldr/Cargo.toml (hunk 2 above)
Comment (Manishearth) [reply provider/cldr/Cargo.toml]: taking the emoji react as approval

**Comment by Manishearth** [comment]
You're clear to merge when this is green

File: provider/cldr/Cargo.toml (hunk 2 above)
Comment (zbraniecki) [reply provider/cldr/Cargo.toml]: I'm comfortable with deps on datagen.

**Pull request merged**